core/src/main/java/org/apache/iceberg/io/PartitionProjection.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructProjection;
+
+public interface PartitionProjection extends StructLike {
+
+  PartitionProjection create(PartitionSpec spec, Types.StructType commonPartitionType);
+
+  PartitionProjection wrap(StructLike newStruct);
+
+  static PartitionProjection of(PartitionSpec spec, Types.StructType commonPartitionType) {
+    if (spec.isUnpartitioned()) {
+      return UnpartitionedSpecProjection.INSTANCE;
+    } else {
+      return new PartitionedSpecProjection().create(spec, commonPartitionType);
+    }
+  }
+
+  class UnpartitionedSpecProjection implements PartitionProjection {
+    static final PartitionProjection INSTANCE = new UnpartitionedSpecProjection();
+
+    @Override
+    public PartitionProjection create(PartitionSpec spec, Types.StructType commonPartitionType) {
+      return this;
+    }
+
+    @Override
+    public PartitionProjection wrap(StructLike newStruct) {
+      return null;
```

Comment:
I traced through the code. this seems to be the relevant fix.

It tries to influence the `OutputFileFactory`, which then calls `LocationProvider` interface.
```
  public EncryptedOutputFile newOutputFile(PartitionSpec spec, StructLike partition) {
    String newDataLocation = locations.newDataLocation(spec, partition, generateFilename());
    OutputFile rawOutputFile = io.newOutputFile(newDataLocation);
    return encryptionManager.encrypt(rawOutputFile);
  }
```

I am wondering if the fix should be done in `OutputFileFactory`. If it is unpartitioned spec, it should call the other `newDataLocation` method without partition from `LocationProvider` interface.
```
  /**
   * Return a fully-qualified data file location for the given filename.
   *
   * @param filename a file name
   * @return a fully-qualified location URI for a data file
   */
  String newDataLocation(String filename);

  /**
   * Return a fully-qualified data file location for the given partition and filename.
   *
   * @param spec a partition spec
   * @param partitionData a tuple of partition data for data in the file, matching the given spec
   * @param filename a file name
   * @return a fully-qualified location URI for a data file
   */
  String newDataLocation(PartitionSpec spec, StructLike partitionData, String filename);
```

Reply:
Basically, I am saying the 2nd `newOutputFile` from `OutputFileFactory` can behave the same as the first method if the partition object is unpartitioned.

```
  /** Generates an {@link EncryptedOutputFile} for unpartitioned writes. */
  public EncryptedOutputFile newOutputFile() {
    OutputFile file = io.newOutputFile(locations.newDataLocation(generateFilename()));
    return encryptionManager.encrypt(file);
  }

  /** Generates an {@link EncryptedOutputFile} for partitioned writes in a given spec. */
  public EncryptedOutputFile newOutputFile(PartitionSpec spec, StructLike partition) {
    String newDataLocation = locations.newDataLocation(spec, partition, generateFilename());
    OutputFile rawOutputFile = io.newOutputFile(newDataLocation);
    return encryptionManager.encrypt(rawOutputFile);
  }
```

Reply:
thanks for the context. I missed the earlier part of the discussion on the first commit/attempt.

Reply:
If we want always null, I think adding `PartitionProjection` is reasonable to avoid if/else in tight loops and to reuse this logic in other places. That said, we know our file builders will always set null for unpartitioned tables. We only need to fix our writers.

What if we fix `RollingFileWriter`? It currently checks `partition == null`.

```
private EncryptedOutputFile newFile() {
  if (partition == null) {
    return fileFactory.newOutputFile();
  } else {
    return fileFactory.newOutputFile(spec, partition);
  }
}
```

We could replace it with `spec.isUnpartitioned()`.

I also don't mind adding an extra branch to `OutputFileFactory` to be safe but I guess that's optional.

Reply:
Fixing the writers seems to be a smaller change, so maybe we can start with that?

Reply:
@aokolnychyi That works for me. Let me update the PR